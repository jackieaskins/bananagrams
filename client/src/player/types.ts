export type Player = {
  userId: string;
  username: string;
  isReady: boolean;
  isTopBanana: boolean;
  isAdmin: boolean;
  gamesWon: number;
};
