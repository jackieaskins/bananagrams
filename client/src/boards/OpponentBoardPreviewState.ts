import { useEffect, useState } from 'react';
import { Player } from '../players/types';

type OpponentBoardPreviewState = {
  handleLeftClick: () => void;
  handleRightClick: () => void;
  handleSelectedPlayerChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  selectedPlayerIndex: number;
  selectedUserId?: string;
};

export const useOpponentBoardPreview = (
  opponents: Player[],
  initialPlayerIndex: number
): OpponentBoardPreviewState => {
  const [selectedUserId, setSelectedUserId] = useState<string | undefined>(
    opponents[initialPlayerIndex]?.userId
  );

  const foundIndex = opponents.findIndex(
    ({ userId }) => userId === selectedUserId
  );
  const selectedPlayerIndex = foundIndex < 0 ? 0 : foundIndex;

  useEffect(() => {
    if (!opponents.some(({ userId }) => userId === selectedUserId)) {
      setSelectedUserId(opponents[0]?.userId);
    }
  }, [opponents, selectedUserId]);

  const handleLeftClick = (): void => {
    const index =
      selectedPlayerIndex <= 0 ? opponents.length - 1 : selectedPlayerIndex - 1;
    setSelectedUserId(opponents[index]?.userId);
  };

  const handleRightClick = (): void => {
    const index =
      selectedPlayerIndex >= opponents.length - 1 ? 0 : selectedPlayerIndex + 1;
    setSelectedUserId(opponents[index]?.userId);
  };

  const handleSelectedPlayerChange = (
    e: React.ChangeEvent<HTMLInputElement>
  ): void => {
    setSelectedUserId(e.target.value);
  };

  return {
    handleLeftClick,
    handleRightClick,
    handleSelectedPlayerChange,
    selectedPlayerIndex,
    selectedUserId,
  };
};
