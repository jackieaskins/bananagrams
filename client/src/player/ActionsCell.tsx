import { DeleteOutlined } from '@ant-design/icons';
import { Button } from 'antd';

import { useCurrentPlayer } from '../game/stateHooks';
import { kickPlayer } from '../socket';
import { Player } from './types';

type ActionsCellProps = {
  player: Player;
};

const ActionsCell = ({
  player: { userId },
}: ActionsCellProps): JSX.Element | null => {
  const currentPlayer = useCurrentPlayer();

  if (!currentPlayer) {
    return null;
  }

  const { isAdmin, userId: currentPlayerId } = currentPlayer;

  if (isAdmin && userId !== currentPlayerId) {
    return (
      <Button
        icon={<DeleteOutlined />}
        onClick={() => {
          kickPlayer({ userId });
        }}
        type="text"
      />
    );
  }

  return null;
};

export default ActionsCell;
