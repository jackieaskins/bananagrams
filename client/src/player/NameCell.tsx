import { CrownOutlined, KeyOutlined } from '@ant-design/icons';
import { Space } from 'antd';

import { Player } from '../players/types';

type NameCellProps = {
  player: Player;
};

const NameCell = ({
  player: { isAdmin, isTopBanana, username },
}: NameCellProps): JSX.Element => (
  <Space>
    <span>{username}</span>
    {isAdmin && (
      <span>
        <KeyOutlined />
      </span>
    )}
    {isTopBanana && (
      <span>
        <CrownOutlined />
      </span>
    )}
  </Space>
);

export default NameCell;
