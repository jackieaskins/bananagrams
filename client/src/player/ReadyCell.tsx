import { CheckOutlined, CloseOutlined } from '@ant-design/icons';
import { Checkbox } from 'antd';

import { getUserId, setReady } from '../socket';
import { Player } from './types';

type ReadyCellProps = {
  player: Player;
};

const ReadyCell = ({
  player: { userId, isReady },
}: ReadyCellProps): JSX.Element => {
  const isCurrentUser = userId === getUserId();

  if (isCurrentUser) {
    return (
      <Checkbox
        checked={isReady}
        onChange={({ target: { checked } }) => {
          setReady({ isReady: checked });
        }}
      />
    );
  }

  if (isReady) {
    return <CheckOutlined css={{ color: 'green' }} />;
  }

  return <CloseOutlined css={{ color: 'red' }} />;
};
export default ReadyCell;
