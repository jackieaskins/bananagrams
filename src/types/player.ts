import { Board } from "./board";
import { Hand } from "./hand";

export enum PlayerStatus {
  NOT_READY = "NOT_READY",
  SPECTATING = "SPECTATING",
  READY = "READY",
}

export type Player = {
  userId: string;
  username: string;
  status: PlayerStatus;
  isTopBanana: boolean;
  isAdmin: boolean;
  gamesWon: number;
  hand: Hand;
  board: Board;
};
