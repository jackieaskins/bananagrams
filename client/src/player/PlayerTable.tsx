import { Table, Typography } from 'antd';
import { ColumnType } from 'antd/lib/table';

import { useCurrentPlayer, useGamePlayers } from '../game/stateHooks';
import ActionsCell from './ActionsCell';
import NameCell from './NameCell';
import ReadyCell from './ReadyCell';
import { Player } from './types';

const COLUMN_ALIGN = 'center';
const MAX_PLAYERS = 8;

const { Text } = Typography;

const actionsColumn: ColumnType<Player> = {
  align: COLUMN_ALIGN,
  key: 'actions',
  render: function ActionsRender(_: any, player: Player) {
    return <ActionsCell player={player} />;
  },
};

const PlayerTable = (): JSX.Element => {
  const players = useGamePlayers();
  const isAdmin = !!useCurrentPlayer()?.isAdmin;

  return (
    <Table
      bordered
      columns={[
        {
          title: 'Players',
          children: [
            {
              align: COLUMN_ALIGN,
              key: 'ready',
              title: 'Ready?',
              render: function ReadyRender(_, player) {
                return <ReadyCell player={player} />;
              },
            },
            {
              align: COLUMN_ALIGN,
              key: 'player',
              title: 'Player',
              render: function PlayerRender(_, player) {
                return <NameCell player={player} />;
              },
            },
            {
              align: COLUMN_ALIGN,
              dataIndex: 'gamesWon',
              title: 'Games won',
            },
            ...(isAdmin && players.length > 1 ? [actionsColumn] : []),
          ],
        },
      ]}
      dataSource={players}
      footer={() => (
        <div css={{ fontSize: '12px', textAlign: 'center' }}>
          <Text type="secondary">
            The game will start when all players are ready
            <br />
            Current player count: {players.length} / {MAX_PLAYERS}
          </Text>
        </div>
      )}
      pagination={false}
      rowKey="userId"
      size="middle"
    />
  );
};

export default PlayerTable;
