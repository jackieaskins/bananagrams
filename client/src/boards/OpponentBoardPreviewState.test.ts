import { useEffect } from 'react';

import { playerFixture } from '../fixtures/player';
import { useOpponentBoardPreview } from './OpponentBoardPreviewState';

const mockSetSelectedUserId = jest.fn();
jest.mock('react', () => ({
  useEffect: jest.fn(),
  useState: jest
    .fn()
    .mockImplementation((initialValue) => [
      initialValue,
      mockSetSelectedUserId,
    ]),
}));

describe('useOpponentBoardPreview', () => {
  it('returns correct default state', () => {
    expect(
      useOpponentBoardPreview([playerFixture({ userId: '123' })], 0)
    ).toMatchSnapshot();
  });

  describe('use effect', () => {
    it('is called whenever opponents or selected user id changes', () => {
      const opponents = [playerFixture()];
      useOpponentBoardPreview(opponents, 0);

      expect(useEffect.mock.calls[0][1]).toEqual([
        opponents,
        opponents[0].userId,
      ]);
    });

    it('updates the selected user id if the current selected user does not exist', () => {
      const opponents = [playerFixture()];
      useOpponentBoardPreview(opponents, 1);

      useEffect.mock.calls[0][0]();

      expect(mockSetSelectedUserId).toHaveBeenCalledWith(opponents[0].userId);
    });

    it('does not update the selected user id if the current selected user exists', () => {
      const opponents = [playerFixture()];
      useOpponentBoardPreview(opponents, 0);

      useEffect.mock.calls[0][0]();

      expect(mockSetSelectedUserId).not.toHaveBeenCalled();
    });

    it('handles no opponents', () => {
      useOpponentBoardPreview([], 0);

      useEffect.mock.calls[0][0]();

      expect(mockSetSelectedUserId).toHaveBeenCalledWith(undefined);
    });
  });

  describe('handleLeftClick', () => {
    const opponents = [playerFixture(), playerFixture()];

    it('decrements the selected index by 1 if possible', () => {
      useOpponentBoardPreview(opponents, 1).handleLeftClick();

      expect(mockSetSelectedUserId).toHaveBeenCalledWith(opponents[0].userId);
    });

    it('moves to end of opponents list if at start', () => {
      useOpponentBoardPreview(opponents, 0).handleLeftClick();

      expect(mockSetSelectedUserId).toHaveBeenCalledWith(opponents[1].userId);
    });

    it('handles no opponents', () => {
      useOpponentBoardPreview([], 0).handleLeftClick();

      expect(mockSetSelectedUserId).toHaveBeenCalledWith(undefined);
    });
  });

  describe('handleRightClick', () => {
    const opponents = [playerFixture(), playerFixture()];

    it('increments the selected index by 1 if possible', () => {
      useOpponentBoardPreview(opponents, 0).handleRightClick();

      expect(mockSetSelectedUserId).toHaveBeenCalledWith(opponents[1].userId);
    });

    it('moves to start of opponents list if at end', () => {
      useOpponentBoardPreview(opponents, 1).handleRightClick();

      expect(mockSetSelectedUserId).toHaveBeenCalledWith(opponents[0].userId);
    });

    it('handles no opponents', () => {
      useOpponentBoardPreview([], 0).handleRightClick();

      expect(mockSetSelectedUserId).toHaveBeenCalledWith(undefined);
    });
  });

  it('handleSelectedPlayerChange sets selected user id', () => {
    const opponents = [playerFixture(), playerFixture()];
    const id = opponents[1].userId;

    useOpponentBoardPreview(opponents, 0).handleSelectedPlayerChange({
      target: { value: id },
    });

    expect(mockSetSelectedUserId).toHaveBeenCalledWith(id);
  });
});
