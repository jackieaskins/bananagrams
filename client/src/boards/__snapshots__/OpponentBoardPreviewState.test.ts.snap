// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useOpponentBoardPreview returns correct default state 1`] = `
{
  "handleLeftClick": [Function],
  "handleRightClick": [Function],
  "handleSelectedPlayerChange": [Function],
  "selectedPlayerIndex": 0,
  "selectedUserId": "123",
}
`;
